rust/chains/hyperlane-sealevel/src/interchain_gas.rs

```diff
@@ -47,12 +97,138 @@ impl InterchainGasPaymaster for SealevelInterchainGasPaymaster {}
 
 /// Struct that retrieves event data for a Sealevel IGP contract
 #[derive(Debug)]
-pub struct SealevelInterchainGasPaymasterIndexer {}
+pub struct SealevelInterchainGasPaymasterIndexer {
+    rpc_client: RpcClientWithDebug,
+    igp: SealevelInterchainGasPaymaster,
+}
+
+/// IGP payment data on Sealevel
+#[derive(Debug, new)]
+pub struct SealevelGasPayment {
+    payment: InterchainGasPayment,
+    log_meta: LogMeta,
+    igp_account_pubkey: H256,
+}
 
 impl SealevelInterchainGasPaymasterIndexer {
     /// Create a new Sealevel IGP indexer.
-    pub fn new(_conf: &ConnectionConf, _locator: ContractLocator) -> Self {
-        Self {}
+    pub async fn new(
+        conf: &ConnectionConf,
+        igp_account_locator: ContractLocator<'_>,
+    ) -> ChainResult<Self> {
+        // Set the `processed` commitment at rpc level
+        let rpc_client = RpcClientWithDebug::new_with_commitment(
+            conf.url.to_string(),
+            CommitmentConfig::processed(),
+        );
+
+        let igp = SealevelInterchainGasPaymaster::new(conf, &igp_account_locator).await?;
+        Ok(Self { rpc_client, igp })
+    }
+
+    async fn get_payment_with_sequence(
+        &self,
+        sequence_number: u64,
+    ) -> ChainResult<SealevelGasPayment> {
+        let payment_bytes = &[
+            &hyperlane_sealevel_igp::accounts::GAS_PAYMENT_DISCRIMINATOR[..],
+            &sequence_number.to_le_bytes()[..],
+        ]
+        .concat();
+        #[allow(deprecated)]
+        let payment_bytes: String = base64::encode(payment_bytes);
+
+        // First, find all accounts with the matching gas payment data.
+        // To keep responses small in case there is ever more than 1
+        // match, we don't request the full account data, and just request
+        // the `unique_gas_payment_pubkey` field.
+        #[allow(deprecated)]
+        let memcmp = RpcFilterType::Memcmp(Memcmp {
```

Comment:
Yeah this logic is already somewhat duplicated with the one in `mailbox.rs`, but since there's just two uses I also think it's fine for now. Created issue https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/2695